modules/dxp/apps/saml/saml-opensaml-integration/src/main/java/com/liferay/saml/opensaml/integration/internal/resolver/DefaultAttributeResolver.java

```diff
@@ -184,17 +184,14 @@ private void _addGroupsAttribute(
 				nameFormat = Attribute.UNSPECIFIED;
 			}
 
-			Stream<Group> groupsStream = groups.stream();
+			List<String> groupsNames = TransformUtil.transform(
+				groups, Group::getName);
 
 			attributePublisher.publish(
 				name, nameFormat,
-				groupsStream.map(
-					Group::getName
-				).map(
-					attributePublisher::buildString
-				).toArray(
-					AttributePublisher.AttributeValue[]::new
-				));
+				TransformUtil.transformToArray(
```

Comment:
@jeffwu0724 Here the 2 `TransformUtil` usages can be combined into 1:

```
			attributePublisher.publish(
				name, nameFormat,
				TransformUtil.transformToArray(
					groups,
					group -> attributePublisher.buildString(group.getName()),
					AttributePublisher.AttributeValue.class));
```

Please review the `TransformUtil` usages and see if the others can also be combined.

If we don't consider converting a list to an array, basically each `TransformUtil` call iterates the passed collection to do the transform; if we call it twice, we have 2 iterations, which reduces the optimization over using stream.

cc @tinatian @slnn @W2jx @JanisZhang @LisztLiszt @julschong